stacker/providers/aws/default.py

```diff
@@ -448,17 +453,29 @@ def generate_stack_policy_args(stack_policy=None):
 
 
 class ProviderBuilder(object):
-    """Implements a ProviderBuilder for the AWS provider."""
+    """Implements a Memoized ProviderBuilder for the AWS provider."""
 
     def __init__(self, region=None, **kwargs):
+        self.lock = Lock()
         self.region = region
         self.kwargs = kwargs
 
     def build(self, region=None, profile=None):
-        if not region:
-            region = self.region
-        session = get_session(region=region, profile=profile)
-        return Provider(session, region=region, **self.kwargs)
+
+        # a thread safe and memoized, provider builder.
+        with self.lock:
+            # memoization lookup key derived by taking md5 of region + profile.
+            key = md5("%s%s" % (region, profile)).hexdigest()
+            
+            # create a new provider on key miss.
+            if key not in self.providers:
```

Comment:
I'd change this - since we'll likely find the key more often than not, it would probably be better as a try/except block:


```
try:
    return self.providers[key]
except KeyError:
    ... build new provider and return it ...
```

Also, self.providers isn't instantiated anywhere that I can tell before this - probably need a `self.providers = {}` in the `__init__` method.